primitives/state-machine/src/lib.rs

```diff
@@ -347,11 +351,11 @@ impl<'a, B, H, N, Exec> StateMachine<'a, B, H, N, Exec> where
 				CallResult<R, Exec::Error>,
 			) -> CallResult<R, Exec::Error>
 	{
-		let pending_changes = self.overlay.clone_pending();
+		self.overlay.start_transaction();
 		let (result, was_native) = self.execute_aux(true, native_call.take());
 
 		if was_native {
-			self.overlay.replace_pending(pending_changes);
+			self.overlay.rollback_transaction().expect(PROOF_CLOSE_TRANSACTION);
 			let (wasm_result, _) = self.execute_aux(
```

Comment:
Nope. We do not intent to roll it back. It matches the semantic of the original code.

Reply:
yes, :+1:  it matches previous, since we added some transaction start in call_api_at.
It is also ok since the rollback will only happen in call_api_at.
Things start to make sense to me (this is actually way better to do things this way considering transaction cost)